main.py

```diff
@@ -0,0 +1,162 @@
+    ("I", "GB-WLS", "cy"): "1001ac43-093d-425c-ac7d-68df5147c603",
+    ("I", "GB-NIR", "en"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("I", "GB-NIR", "ga"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("I", "GB-NIR", "eo"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("C", "GB-ENG", "en"): "4077d2cf-81cd-462d-9065-f227a7c39a8d",
+    ("C", "GB-WLS", "en"): "4077d2cf-81cd-462d-9065-f227a7c39a8d",
+    ("C", "GB-WLS", "cy"): "e4a4ebea-fcc8-463b-8686-5b8a7320f089",
+}
+
+data_fields = ("email_address", "display_address", "tx_id", "questionnaire_id")
+
+
+session = Session()
+
+
+def _validate_request(request: Request) -> Tuple[Mapping, str, Mapping]:
+    if not request.method == "POST":
+        # Note that Cloud Functions expect serialized JSON
+        # to correctly log
+        # https://cloud.google.com/functions/docs/monitoring/logging#writing_structured_logs
+        raise InvalidRequestError("Method not allowed", 405)
+
+    if not request.json:
+        raise InvalidRequestError("Missing notification request data", 422)
+
+    data = request.json["payload"]["fulfilmentRequest"]
+
+    log_context = {
+        "tx_id": data.get("tx_id"),
+        "questionnaire_id": data.get("questionnaire_id"),
+    }
+
+    required_keys = ("form_type", "region_code", "language_code")
+    if missing_keys := [key for key in required_keys if not data.get(key)]:
+        msg = f"Missing {', '.join(missing_keys)} identifier(s)"
+        raise InvalidRequestError(msg, 422, log_context)
+
+    try:
+        template_id = template_id_mapping[
+            (data["form_type"], data["region_code"], data["language_code"])
+        ]
+    except KeyError as exc:
+        raise InvalidRequestError("No template id selected", 422, log_context) from exc
+
+    return {key: data.get(key) for key in data_fields}, template_id, log_context
+
+
+# pylint: disable=too-many-return-statements
+def send_email(request: Request) -> Tuple[str, int]:
+    try:
+        data, template_id, log_context = _validate_request(request)
+    except InvalidRequestError as error:
+        log_context = error.log_context if error.log_context else {}
+        log_error(message=error.message, **log_context)
+        return error.message, error.status_code
+
+    api_token = create_jwt_token(secret_key, service_id)
+    session.headers.update(
+        {
+            "Content-type": "application/json",
+            "Authorization": f"Bearer {api_token}",
+        }
+    )
+
+    kwargs = json.dumps(
+        {
+            "template_id": template_id,
+            "personalisation": {"address": data["display_address"]},
+            "email_address": data["email_address"],
+        }
+    )
+
+    try:
+        response = session.post(
+            f"{NOTIFY_BASE_URL}/notifications/email",
+            kwargs,
+        )
+        response.raise_for_status()
+        entry = dict(**log_context, message="notify email requested")
+        log_info(**entry)
```

Comment:
I don't think there is a need for the `entry` variable, and we should call the log method consistently.
```suggestion
        log_info(message="notify email requested", **log_context)
```

Reply:
Changed as part of the other suggestions (making `message` a positional arg)